aibolit/metrics/lcom4/lcom4.py

```diff
@@ -20,61 +20,255 @@
 # OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 # SOFTWARE.
 
-# import matplotlib.pyplot as plt
 import networkx as nx  # type: ignore
-from collections import defaultdict
+from networkx import Graph
 from aibolit.utils.ast import AST
-from typing import Set, Dict, List
+from typing import List, Generator, Tuple, Any, Union, overload
 from javalang.tree import ClassDeclaration, InterfaceDeclaration, MethodDeclaration, \
-    MemberReference, FieldDeclaration, MethodInvocation, This, Node
+    MemberReference, FieldDeclaration, MethodInvocation, This, Node, LocalVariableDeclaration
+FldExh = Tuple[str, Tuple[str, str]]
+MthExh = Tuple[str, Tuple[Tuple[str, str], ...]]
+Nodes = Tuple[tuple, Node]
+NodeGen = Generator[Tuple[tuple, Node], None, None]
+EdgeNode = Union[MthExh, FldExh]
+JLang = Union[ClassDeclaration, InterfaceDeclaration, MethodDeclaration,
+              MemberReference, FieldDeclaration, MethodInvocation, This, Node, LocalVariableDeclaration]
 
 
 class LCOM4:
 
     def value(self, filename: str) -> int:
         tree: Node = AST(filename).value()
-        graph: Dict[str, Set[str]] = defaultdict(set)
+        G: Graph = nx.Graph()
 
-        fields: List[str] = [node.declarators[0].name for _, node in tree.filter(FieldDeclaration)]
-        methods: List[str] = [node.name for _, node in tree.filter(MethodDeclaration)]
-        interfaces: List[InterfaceDeclaration] = [node for _, node in tree.filter(InterfaceDeclaration)]
-        current_class: ClassDeclaration = list(tree.filter(ClassDeclaration))[0][1]
-        interfaces_methods: Set[str] = set()
-        nested_methods: Set[str] = set()
+        for class_path, class_node in tree.filter(ClassDeclaration):
 
-        class_decl: List[ClassDeclaration] = \
-            [node for _, node in current_class.filter(ClassDeclaration) if node.name != current_class.name]
+            # Extract all methods and fields
+            # from ClassDeclaration node
 
-        for i in interfaces:
-            interfaces_methods.update([node.name for _, node in i.filter(MethodDeclaration)])
-        for k in class_decl:
-            nested_methods.update([node.name for _, node in k.filter(MethodDeclaration)])
+            field_nodes: List[Nodes] = list(self.filter_node_lvl(class_node, FieldDeclaration))
```

Comment:
why you return Generator if you always convert it to the list?

Reply:
This is made for the opportunity to iterate trough the generator itself if needed.

Reply:
Resolved